File: backend/utils/watermark.py
"""
Watermark utility for adding watermarks to free tier diagram exports.
"""
from PIL import Image, ImageDraw, ImageFont
from io import BytesIO
import os


def add_watermark_to_image(
    image_bytes: bytes,
    watermark_text: str = "Created with ArchitectAI - Free Plan",
    opacity: float = 0.3,
    position: str = "bottom-right"
) -> bytes:
    """
    Add a watermark to an image.
    
    Args:
        image_bytes: Original image as bytes
        watermark_text: Text to use as watermark
        opacity: Watermark opacity (0.0 to 1.0)
        position: Watermark position ("bottom-right", "bottom-left", "center")
    
    Returns:
        Watermarked image as bytes
    """
    # Open image
    image = Image.open(BytesIO(image_bytes))
    
    # Convert to RGBA if needed
    if image.mode != "RGBA":
        image = image.convert("RGBA")
    
    # Create watermark layer
    watermark = Image.new("RGBA", image.size, (0, 0, 0, 0))
    draw = ImageDraw.Draw(watermark)
    
    # Try to use a nice font, fallback to default
    try:
        font_size = int(image.height / 30)  # Adaptive font size
        font = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", font_size)
    except:
        font = ImageFont.load_default()
    
    # Get text bounding box
    bbox = draw.textbbox((0, 0), watermark_text, font=font)
    text_width = bbox[2] - bbox[0]
    text_height = bbox[3] - bbox[1]
    
    # Calculate position
    margin = 20
    if position == "bottom-right":
        x = image.width - text_width - margin
        y = image.height - text_height - margin
    elif position == "bottom-left":
        x = margin
        y = image.height - text_height - margin
    elif position == "center":
        x = (image.width - text_width) // 2
        y = (image.height - text_height) // 2
    else:
        x = image.width - text_width - margin
        y = image.height - text_height - margin
    
    # Calculate opacity (0-255)
    alpha = int(255 * opacity)
    
    # Draw watermark
    draw.text(
        (x, y),
        watermark_text,
        fill=(128, 128, 128, alpha),  # Gray with opacity
        font=font
    )
    
    # Composite watermark onto image
    watermarked = Image.alpha_composite(image, watermark)
    
    # Convert back to RGB for JPEG/PNG
    if watermarked.mode == "RGBA":
        watermarked = watermarked.convert("RGB")
    
    # Save to bytes
    output = BytesIO()
    watermarked.save(output, format="PNG")
    return output.getvalue()


def should_add_watermark(user_plan: str) -> bool:
    """
    Check if watermark should be added based on user plan.
    
    Args:
        user_plan: User's subscription plan
    
    Returns:
        True if watermark should be added, False otherwise
    """
    return user_plan.lower() == "free"

File: frontend/lib/watermark.ts
/**
 * Add watermark to an image using HTML5 Canvas.
 * @param imageBlob - Original image blob
 * @param watermarkText - Text to display as watermark
 * @returns Promise<Blob> - Watermarked image blob
 */
export async function addWatermarkToImage(
    imageBlob: Blob,
    watermarkText: string = "Created with ArchitectAI - Free Plan"
): Promise<Blob> {
    return new Promise((resolve, reject) => {
        const img = new Image();
        img.crossOrigin = "anonymous"; // Prevent canvas tainting
        const url = URL.createObjectURL(imageBlob);

        img.onload = () => {
            // Create canvas
            const canvas = document.createElement("canvas");
            canvas.width = img.width;
            canvas.height = img.height;
            const ctx = canvas.getContext("2d");

            if (!ctx) {
                reject(new Error("Could not get canvas context"));
                return;
            }

            // Draw original image
            ctx.drawImage(img, 0, 0);

            // Add watermark
            const fontSize = Math.max(img.height / 30, 14); // Adaptive font size
            ctx.font = `${fontSize}px Arial, sans-serif`;
            ctx.fillStyle = "rgba(128, 128, 128, 0.3)"; // Gray with 30% opacity
            ctx.textAlign = "right";
            ctx.textBaseline = "bottom";

            // Position at bottom-right corner with margin
            const margin = 20;
            const x = img.width - margin;
            const y = img.height - margin;

            ctx.fillText(watermarkText, x, y);

            // Convert canvas to blob
            canvas.toBlob(
                (blob) => {
                    URL.revokeObjectURL(url);
                    if (blob) {
                        resolve(blob);
                    } else {
                        reject(new Error("Failed to create watermarked image"));
                    }
                },
                "image/png",
                1.0
            );
        };

        img.onerror = () => {
            URL.revokeObjectURL(url);
            reject(new Error("Failed to load image"));
        };

        img.src = url;
    });
}

/**
 * Check if user is on free plan and should have watermarked exports.
 */
export function shouldAddWatermark(userPlan?: string): boolean {
    return !userPlan || userPlan.toLowerCase() === "free";
}
